xc/device.ml

```diff
@@ -2059,30 +1885,334 @@ module Dm = struct
+    | Profile.Qemu_upstream_compat -> (module Qemu_upstream_compat : Intf)
+    | Profile.Qemu_upstream        -> (module Qemu_upstream        : Intf)
+
+  let of_domid x = of_profile (Profile.of_domid x)
+end
+
+(*
+ *  Functions using the backend dispatcher
+ *)
+module Vbd = struct
+  include Vbd_Common
+
+  let media_eject ~xs device =
+    let module Q = (val Backend.of_domid device.frontend.domid) in
+    Q.Vbd.qemu_media_change ~xs device "" ""
+
+  let media_insert ~xs ~phystype ~params device =
+    let _type = backendty_of_physty phystype in
+    let module Q = (val Backend.of_domid device.frontend.domid) in
+    Q.Vbd.qemu_media_change ~xs device _type params
+
+end (* Vbd *)
+
+module Dm = struct
+  include Dm_Common
+
+  let get_vnc_port ~xs domid =
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.get_vnc_port ~xs domid
+
+  let init_daemon ~task ~path ~args ~name ~domid ~xs ~ready_path ?ready_val ~timeout ~cancel _ =
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.init_daemon ~task ~path ~args ~name ~domid ~xs ~ready_path ?ready_val ~timeout ~cancel ()
+
+  let suspend (task: Xenops_task.task_handle) ~xs ~qemu_domid domid =
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.suspend task ~xs ~qemu_domid domid
+
+  let stop ~xs ~qemu_domid domid  =
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.stop ~xs ~qemu_domid domid
 
   let maybe_write_pv_feature_flags ~xs domid =
-    if is_upstream_qemu domid then
-      match is_hvm_linux domid with
-      | Result.Ok true ->
-        let write_local_domain prefix x = xs.Xs.write (Printf.sprintf "/local/domain/%d/%s%s" domid prefix x) "1" in
-        List.iter (write_local_domain "control/feature-") ["suspend"; "shutdown"; "vcpu-hotplug"];
-        List.iter (write_local_domain "data/") ["updated"]
-      | _ -> ()
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.maybe_write_pv_feature_flags ~xs domid
+
+  let with_dirty_log domid ~f =
+    let module Q = (val Backend.of_domid domid) in
+    Q.Dm.with_dirty_log domid ~f
+
+end (* Dm *)
 
-end (* End of module Dm *)
+(* functions depending on modules Vif, Vbd, Pci, Vfs, Vfb, Vkbd, Dm *)
+
+let hard_shutdown (task: Xenops_task.task_handle) ~xs (x: device) = match x.backend.kind with
+  | Vif -> Vif.hard_shutdown task ~xs x
+  | Vbd _ | Tap -> Vbd.hard_shutdown task ~xs x
+  | Pci -> PCI.hard_shutdown task ~xs x
+  | Vfs -> Vfs.hard_shutdown task ~xs x
+  | Vfb -> Vfb.hard_shutdown task ~xs x
+  | Vkbd -> Vkbd.hard_shutdown task ~xs x
+
+let clean_shutdown (task: Xenops_task.task_handle) ~xs (x: device) = match x.backend.kind with
+  | Vif -> Vif.clean_shutdown task ~xs x
+  | Vbd _ | Tap -> Vbd.clean_shutdown task ~xs x
+  | Pci -> PCI.clean_shutdown task ~xs x
+  | Vfs -> Vfs.clean_shutdown task ~xs x
+  | Vfb -> Vfb.clean_shutdown task ~xs x
+  | Vkbd -> Vkbd.clean_shutdown task ~xs x
 
 let get_vnc_port ~xs domid = 
```

Comment:
We have a few functions with the same name and different meanings, so it will be nice as a future PR to refactor all these to be more coherent.